pilot/pkg/model/telemetry_test.go

```diff
@@ -0,0 +1,301 @@
+						Providers: []*tpb.ProviderRef{
+							{
+								Name: "zipkin",
+							},
+						},
+						RandomSamplingPercentage: 10.10,
+					},
+				},
+			},
+		},
+		{
+			name: "workload-agnostic in namespace and no workload labels",
+			ns:   "foo",
+			configs: []config.Config{
+				newTelemetry("root", "istio-system", rootTrace),
+				newTelemetry("foo", "foo", barTrace),
+			},
+			want: &tpb.Telemetry{
+				Tracing: []*tpb.Tracing{
+					{
+						Providers: []*tpb.ProviderRef{
+							{
+								Name: "zipkin",
+							},
+						},
+						RandomSamplingPercentage: 10.10,
+						DisableSpanReporting:     true,
+					},
+				},
+			},
+		},
+		{
+			name:           "both workload-specific and workload-agnostic in namespace with workload labels",
+			ns:             "foo",
+			workloadLabels: map[string]string{"service.istio.io/canonical-name": "foo"},
+			configs: []config.Config{
+				newTelemetry("root", "istio-system", rootTrace),
+				newTelemetry("foo", "foo", fooTrace),
+				newTelemetry("bar", "foo", barTrace),
+				newTelemetry("baz", "baz", bazTrace),
+			},
+			want: &tpb.Telemetry{
+				Tracing: []*tpb.Tracing{
+					{
+						Providers: []*tpb.ProviderRef{
+							{
+								Name: "zipkin",
+							},
+						},
+						RandomSamplingPercentage: 77.77,
+						// no disableSpanReporting because the missing value overrides
+					},
+				},
+			},
+		},
+		{
+			name: "both workload-specific and workload-agnostic in namespace without workload labels",
+			ns:   "foobar",
+			configs: []config.Config{
+				newTelemetry("root", "istio-system", rootTrace),
+				newTelemetry("foo", "foobar", fooTrace),
+				newTelemetry("bar", "foobar", barTrace),
+				newTelemetry("baz", "baz", bazTrace),
+			},
+			want: &tpb.Telemetry{
+				Tracing: []*tpb.Tracing{
+					{
+						Providers: []*tpb.ProviderRef{
+							{
+								Name: "zipkin",
+							},
+						},
+						RandomSamplingPercentage: 10.10,
+						DisableSpanReporting:     true,
+					},
+				},
+			},
+		},
+		{
+			name: "provider and custom tags override",
+			ns:   "baz",
+			configs: []config.Config{
+				newTelemetry("root", "istio-system", rootTrace),
+				newTelemetry("foo", "foo", fooTrace),
+				newTelemetry("baz", "baz", bazTrace),
+			},
+			want: &tpb.Telemetry{
+				Tracing: []*tpb.Tracing{
+					{
+						Providers: []*tpb.ProviderRef{
+							{
+								Name: "not-zipkin",
+							},
+						},
+						RandomSamplingPercentage: 10.10,
+						CustomTags: map[string]*tpb.Tracing_CustomTag{
+							"fun": {Type: &tpb.Tracing_CustomTag_Environment{
+								Environment: &tpb.Tracing_Environment{
+									Name:         "SOME_ENV_VAR",
+									DefaultValue: "EMPTY",
+								},
+							}},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for _, v := range cases {
+		t.Run(v.name, func(tt *testing.T) {
+			telemetries := createTestTelemetries(v.configs, tt)
+			got := telemetries.EffectiveTelemetry(v.ns, []labels.Instance{v.workloadLabels})
+			if !reflect.DeepEqual(got, v.want) {
```

Comment:
Does this work? I thought you had to do `protocmp.Transform` to compare protos?

Reply:
Ah nevermind - that is only for `golang/protobuf` - we use `gogo`

Reply:
switched anyway.